projects/odf_transform/odf_transform/odf_parser/odf_parser.py

```diff
@@ -199,34 +190,100 @@ def _find_previous_key(key_list, present_key):
                 previous_key = i
         return previous_key
 
-    # Find matching vocabulary
+    # Loop through each variables and detect flag variables
+    for var, att in metadata.items():
+        # Retrieve information from variable name
+        odf_var_name = parse_odf_code_variable(var)
+        # FLAG VARIABLES Detect if it is a flag column and find its related variable
+        flag_column = False
+        related_variable = None
+        if odf_var_name['name'] == 'QQQQ':  # MLI FLAG should apply to previous variable
+            related_variable = _find_previous_key(metadata, var)
+            flag_column = True
+        elif odf_var_name['name'] in ['QCFF', 'FFFF']:
+            flag_column = True
+        elif var.startswith('Q') and var[1:] in metadata.keys():
+            # BIO Format which Q+[PCODE] of the associated variable
+            related_variable = var[1:]
+            flag_column = True
+
+        # Make sure that the flag column relates to a specific variable and try to make sure it's the right match.
+        if flag_column and var not in ['QCFF', 'FFFF']:
+            if related_variable and related_variable not in metadata:
+                warnings.warn('{0} flag is referring to {1} which is not available as variable'
+                              .format(var, related_variable), UserWarning)
+
+            # Drop odf flag name prefix
+            flag_name_with_no_prefix = re.sub(r"quality\sflag.*:\s*|quality flag of ", '',
+                                              att['original_NAME'], 1, re.IGNORECASE)
+            # Flag name do not match either variable name or code, give a warning.
+            if related_variable and \
+                    flag_name_with_no_prefix not in metadata[related_variable].get('original_NAME') and \
+                    flag_name_with_no_prefix not in metadata[related_variable].get('original_CODE'):
+                warnings.warn(
+                    '{0}[{4}] flag was matched to referring to {1} but odf variable name[{2}] or code[{3}] do not match'
+                        .format(var,
+                                related_variable,
+                                metadata[related_variable].get('original_NAME'),
+                                metadata[related_variable].get('original_CODE'),
+                                att['original_NAME']),
+                    UserWarning)
+
+        # Standardize Flag variable attributes, related variable and add convention attributes
+        if related_variable:
```

Comment:
looks like this runs whether or not its a flag column, is that how it's supposed to work?

Reply:
lots have change here